// @enzymefinance/hardhat/jest/config/environment/dist/enzymefinance-hardhat-jest-config-environment.esm.js
import _asyncToGenerator from '@babel/runtime/helpers/esm/asyncToGenerator';
import _classCallCheck from '@babel/runtime/helpers/esm/classCallCheck';
import _createClass from '@babel/runtime/helpers/esm/createClass';
import _assertThisInitialized from '@babel/runtime/helpers/esm/assertThisInitialized';
import _get from '@babel/runtime/helpers/esm/get';
import _inherits from '@babel/runtime/helpers/esm/inherits';
import _possibleConstructorReturn from '@babel/runtime/helpers/esm/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime/helpers/esm/getPrototypeOf';
import _defineProperty from '@babel/runtime/helpers/esm/defineProperty';
import _regeneratorRuntime from '@babel/runtime/regenerator';
import { createCoverageCollector } from '@enzymefinance/coverage';
import deepmerge from 'deepmerge';
import fs from 'fs-extra';
import { HARDHAT_NETWORK_NAME } from 'hardhat/internal/constants';
import { HardhatContext } from 'hardhat/internal/context';
import { loadConfigAndTasks } from 'hardhat/internal/core/config/config-loading';
import { getEnvHardhatArguments } from 'hardhat/internal/core/params/env-variables';
import { HARDHAT_PARAM_DEFINITIONS } from 'hardhat/internal/core/params/hardhat-params';
import { Environment } from 'hardhat/internal/core/runtime-environment';
import { willRunWithTypescript, loadTsNode } from 'hardhat/internal/core/typescript-support';
import NodeEnvironment from 'jest-environment-node';
import path from 'path';
import { v4 } from 'uuid';
import { E as EthereumTestnetProvider } from '../../../../dist/provider-caf7297f.esm.js';
import 'ethers';
import 'hardhat-deploy-ethers/dist/src/signer-with-address';

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var EnzymeHardhatEnvironment = /*#__PURE__*/function (_NodeEnvironment) {
  _inherits(EnzymeHardhatEnvironment, _NodeEnvironment);

  var _super = _createSuper(EnzymeHardhatEnvironment);

  function EnzymeHardhatEnvironment(config) {
    var _config$coverage, _config$history, _process$env$__HARDHA;

    var _this;

    _classCallCheck(this, EnzymeHardhatEnvironment);

    _this = _super.call(this, config);

    _defineProperty(_assertThisInitialized(_this), "metadataFilePath", '');

    _defineProperty(_assertThisInitialized(_this), "tempDir", '');

    _defineProperty(_assertThisInitialized(_this), "codeCoverageRuntimeRecording", {});

    _defineProperty(_assertThisInitialized(_this), "recordCodeCoverage", false);

    _defineProperty(_assertThisInitialized(_this), "recordCallHistory", true);

    _defineProperty(_assertThisInitialized(_this), "runtimeEnvironment", void 0);

    _defineProperty(_assertThisInitialized(_this), "removeCallHistoryListener", void 0);

    _defineProperty(_assertThisInitialized(_this), "removeCodeCoverageListener", void 0);

    _this.recordCodeCoverage = !!((_config$coverage = config.coverage) !== null && _config$coverage !== void 0 ? _config$coverage : false);
    _this.recordCallHistory = !!((_config$history = config.history) !== null && _config$history !== void 0 ? _config$history : true);
    _this.tempDir = (_process$env$__HARDHA = process.env.__HARDHAT_COVERAGE_TEMPDIR__) !== null && _process$env$__HARDHA !== void 0 ? _process$env$__HARDHA : '';

    if (_this.recordCodeCoverage && !_this.tempDir) {
      throw new Error('Missing shared temporary directory for code coverage data collection');
    }

    _this.runtimeEnvironment = getRuntimeEnvironment();
    _this.metadataFilePath = path.join(_this.runtimeEnvironment.config.codeCoverage.path, 'metadata.json');
    return _this;
  }

  _createClass(EnzymeHardhatEnvironment, [{
    key: "setup",
    value: function () {
      var _setup = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee() {
        var env, provider, metadata, collector;
        return _regeneratorRuntime.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _context.next = 2;
                return _get(_getPrototypeOf(EnzymeHardhatEnvironment.prototype), "setup", this).call(this);

              case 2:
                env = this.runtimeEnvironment;
                provider = new EthereumTestnetProvider(env);
                this.global.hre = env;
                this.global.provider = provider; // Re-route call history recording to whatever is the currently
                // active history object. Required for making history and snapshoting
                // work nicely together.

                if (this.recordCallHistory) {
                  this.removeCallHistoryListener = addListener(env.network.provider, 'beforeMessage', function (message) {
                    provider.history.record(message);
                  });
                }

                if (!this.recordCodeCoverage) {
                  _context.next = 13;
                  break;
                }

                _context.next = 10;
                return fs.readJson(this.metadataFilePath);

              case 10:
                metadata = _context.sent;
                collector = createCoverageCollector(metadata, this.codeCoverageRuntimeRecording);
                this.removeCodeCoverageListener = addListener(env.network.provider, 'step', collector);

              case 13:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function setup() {
        return _setup.apply(this, arguments);
      }

      return setup;
    }()
  }, {
    key: "teardown",
    value: function () {
      var _teardown = _asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee2() {
        var _this$removeCodeCover, _this$removeCallHisto;

        var file, output;
        return _regeneratorRuntime.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                (_this$removeCodeCover = this.removeCodeCoverageListener) === null || _this$removeCodeCover === void 0 ? void 0 : _this$removeCodeCover.call(this);
                (_this$removeCallHisto = this.removeCallHistoryListener) === null || _this$removeCallHisto === void 0 ? void 0 : _this$removeCallHisto.call(this);

                if (!(this.recordCodeCoverage && Object.keys(this.codeCoverageRuntimeRecording).length)) {
                  _context2.next = 7;
                  break;
                }

                file = path.join(this.tempDir, "".concat(v4(), ".json"));
                output = {
                  hits: this.codeCoverageRuntimeRecording,
                  metadata: this.metadataFilePath
                };
                _context2.next = 7;
                return fs.outputJson(file, output, {
                  spaces: 2
                });

              case 7:
                _context2.next = 9;
                return _get(_getPrototypeOf(EnzymeHardhatEnvironment.prototype), "teardown", this).call(this);

              case 9:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function teardown() {
        return _teardown.apply(this, arguments);
      }

      return teardown;
    }()
  }]);

  return EnzymeHardhatEnvironment;
}(NodeEnvironment);
var environment;
function getRuntimeEnvironment() {
  if (environment != null) {
    return environment;
  }

  if (HardhatContext.isCreated()) {
    HardhatContext.deleteHardhatContext();
  }

  var context = HardhatContext.createHardhatContext();
  var args = deepmerge(getEnvHardhatArguments(HARDHAT_PARAM_DEFINITIONS, process.env), {
    emoji: false,
    help: false,
    network: HARDHAT_NETWORK_NAME,
    version: false
  });

  if (willRunWithTypescript(args.config)) {
    loadTsNode();
  }

  var config = loadConfigAndTasks();
  var extenders = context.extendersManager.getExtenders();
  environment = new Environment(config, args, {}, extenders);
  context.setHardhatRuntimeEnvironment(environment);
  return environment;
}
function addListener(provider, event, handler) {
  var inner = provider._provider;

  while (inner._wrapped) {
    inner = inner._wrapped;
  }

  var init = inner._init.bind(inner);

  var subscribed = false;
  var removed = false;
  inner._init = /*#__PURE__*/_asyncToGenerator( /*#__PURE__*/_regeneratorRuntime.mark(function _callee3() {
    var vm;
    return _regeneratorRuntime.wrap(function _callee3$(_context3) {
      while (1) {
        switch (_context3.prev = _context3.next) {
          case 0:
            _context3.next = 2;
            return init();

          case 2:
            if (!subscribed && !removed) {
              subscribed = true;
              vm = inner._node._vm;
              vm.on(event, handler);
            }

          case 3:
          case "end":
            return _context3.stop();
        }
      }
    }, _callee3);
  }));
  return function () {
    var _node;

    if (removed) {
      return;
    }

    removed = true;
    var vm = (_node = inner._node) === null || _node === void 0 ? void 0 : _node._vm;

    if (vm != null) {
      vm.off(event, handler);
    }
  };
}

export default EnzymeHardhatEnvironment;
export { addListener, getRuntimeEnvironment };
